tests/test_settings.py

```diff
@@ -128,6 +130,62 @@ def test_base_path_overrides_config_file_code_and_deps(
     assert settings.deps == expected
 
 
+OPTION_VALUES = {
+    "code": ["a", "b", "c"],
+    "deps": ["d", "e"],
+    "ignore-undeclared": ["f", "g"],
+    "ignore-unused": ["h", "i", "j"],
+}
+
+
+def multivalued_optargs_grid() -> List[List[str]]:
+    """Create command-line option/argument combinations which
+    mix order and number of multivalued parameters."""
+
+    def powerset_non_empties(iterable):
+        xs = list(iterable)
+        return chain.from_iterable(combinations(xs, k) for k in range(1, len(xs)))
+
+    opt_by_arg = {arg: opt for opt, argvals in OPTION_VALUES.items() for arg in argvals}
+    parameter_grids = []
+    for items in OPTION_VALUES.values():
+        split_2 = set(
+            chain.from_iterable(
+                [(x, y), (y, x)]
+                for x, y in combinations(powerset_non_empties(items), 2)
+                if len(x + y) == len(items) and not set(x) & set(y)
+            )
+        )
+        parameter_grids.append(split_2)
+    parameter_grids = list(product(*parameter_grids))
+    command_lines = []
+    for param_grid in parameter_grids:
+        xss = list(chain(*param_grid))
+        random.shuffle(xss)
+        optargs = list(chain(*[[f"--{opt_by_arg[xs[0]]}"] + list(xs) for xs in xss]))
+        command_lines.append(optargs)
+    return command_lines
+
+
+@pytest.mark.parametrize("optargs", multivalued_optargs_grid())
+def test_multivalued_options_are_aggregated_correctly(optargs):
+    settings = run_build_settings(optargs)
+    assert settings.code == to_path_set(OPTION_VALUES["code"])
+    assert settings.deps == to_path_set(OPTION_VALUES["deps"])
+    assert settings.ignore_undeclared == set(OPTION_VALUES["ignore-undeclared"])
+    assert settings.ignore_unused == set(OPTION_VALUES["ignore-unused"])
+
+
+@pytest.mark.parametrize(
+    "optname",
+    [act.dest for act in build_parser()._actions if isinstance(act, ArgparseUnionAction)], # pylint: disable=protected-access
+)
+def test_multivalued_options_are_absent_if_not_provided_at_cli(optname):
+    parsed_cli = build_parser().parse_args([])
+    with pytest.raises(AttributeError):
+        getattr(parsed_cli, optname)
```

Comment:
Love this, although I don't see why this need to be parametrized? The code we're _testing_ (`build_parser().parse_args([])`) is the same in all cases, we're just testing multiple (non-)attributes of the returned args. Suggestion:

```suggestion
def test_multivalued_options_are_absent_if_not_provided_at_cli():
    parsed_args = build_parser().parse_args([])
    suppressed_attrs = [
        act.dest
        for act in build_parser()._actions  # pylint: disable=protected-access
        if isinstance(act, ArgparseUnionAction)
    ]
    for attr in suppressed_attrs:
        assert not hasattr(parsed_args, attr)
```

Maybe even consider extending the test to also cover other `Settings` members that correspond directly to `Namespace` members returned from `parse_args()`. After all, the cascade of CLI -> environment -> config file -> hardcoded defaults _depend_ very much on CLI options _not_ being present in this `Namespace` when they're not passed at the command line.

Reply:
I see your point in favor of parametrization. Happy to see where this goes. 😄

Reply:
see what you think here @jherland , my pref isn't super strong here: 400e34a7330944499435ce1a63439179a68d6ca2